packages/caravan-fees/src/rbf.ts

```diff
@@ -0,0 +1,303 @@
+      (sum, utxo) => sum.plus(utxo ? new BigNumber(utxo.split(",")[0]) : 0),
+      new BigNumber(0)
+    );
+  }
+
+  private calculateTotalOutputValue(psbt: PsbtV2): BigNumber {
+    return psbt.PSBT_OUT_AMOUNT.reduce(
+      (sum, amount) => sum.plus(new BigNumber(amount)),
+      new BigNumber(0)
+    );
+  }
+
+  private estimateVsize(psbt: PsbtV2): number {
+    return (
+      psbt.PSBT_GLOBAL_INPUT_COUNT * 148 +
+      psbt.PSBT_GLOBAL_OUTPUT_COUNT * 34 +
+      10
+    );
+  }
+
+  private subtractFeeFromOutputs(psbt: PsbtV2, feeIncrease: BigNumber): void {
+    const nonChangeOutputs = psbt.PSBT_OUT_AMOUNT.filter(
+      (_, index) => !this.isChangeOutput(index, psbt)
+    );
+    const feePerOutput = feeIncrease
+      .dividedBy(nonChangeOutputs.length)
+      .integerValue(BigNumber.ROUND_CEIL);
+    // NEED TO FIX THIS
+    psbt.PSBT_OUT_AMOUNT = psbt.PSBT_OUT_AMOUNT.map((amount, index) => {
+      if (!this.isChangeOutput(index, psbt)) {
+        const newAmount = new BigNumber(amount).minus(feePerOutput);
+        if (newAmount.isLessThan(this.dustThreshold)) {
+          throw new Error("Output would become dust after fee subtraction");
+        }
+        return newAmount.toString();
+      }
+      return amount;
+    });
+  }
+
+  private isChangeOutput(index: number, psbt: PsbtV2): boolean {
+    const bip32Derivation = psbt.PSBT_OUT_BIP32_DERIVATION[index];
+    return bip32Derivation !== null && bip32Derivation.length > 0;
+  }
+}
+
+/**
+ * Prepare an RBF (Replace-By-Fee) transaction.
+ *
+ * "Preparing" a transaction in this context means:
+ * 1. Creating a new transaction based on the original one.
+ * 2. Adjusting the inputs and outputs to accommodate the new, higher fee.
+ * 3. Ensuring the new transaction adheres to RBF rules (BIP 125).
+ * 4. Handling dust outputs and change addresses appropriately.
+ * 5. Calculating and setting the new fee rate.
+ *
+ * The prepared transaction is not yet signed or broadcast. It's a new PSBT
+ * (Partially Signed Bitcoin Transaction) that's ready for signing by the
+ * appropriate parties and subsequent broadcasting to the Bitcoin network.
+ *
+ * This preparation step is crucial for RBF as it allows users to create a
+ * valid replacement transaction with a higher fee, which can then be used
+ * to accelerate confirmation of a stuck transaction or change its outputs
+ * before it's mined into a block.
+ *
+ * @param psbt - The original transaction as a PsbtV2
+ * @param newFeeRate - The new fee rate for the replacement transaction
+ * @param options - Options for preparing the RBF transaction
+ *   - subtractFeeFromOutput: The index of the output to subtract the fee increase from.
+ *If not provided, the fee is subtracted proportionally from all non-change outputs.
+ *   - dustThreshold: The minimum amount (in satoshis) for an output to not be considered dust.
+ * @returns A new PsbtV2 ready for signing and broadcasting
+ */
+export function prepareRbfTransaction(
```

Comment:
Nice, adding utility functions in case people prefer the functional approach is a nice convenience.

Reply:
Should we keep these in the rbf file itself or have another utility file where we can define all these functions for both RBF and CPFP ?